includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
+          return $json;
+        } else {
+          $messages = array(
+            'message' => t("For object @pid : JSON key from $endpoint ($expectation) missing . $response"),
+            'message_sub' => array('@pid' => $object->id),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+        }
+      } else {
+        $messages = array(
+          'message' => t("For object @pid : Expected JSON key from $endpoint but could not decode. $response"),
+          'message_sub' => array('@pid' => $object->id),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+      }
+    } else {
+      $messages = array(
+        'message' => t("For object @pid : Expected one of " .implode(',', array_keys($expect))."; got a $response_code"),
+        'message_sub' => array('@pid' => $object->id), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+  }
+  return;
+}
+
+/*
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param $upload_target - url of upload target
+ * @param $file_path - path of object
+ */
+function islandora_video_streaming_panopto_upload_file($upload_target, $file_path) {
+  $element = explode("/", $upload_target);
+  $prefix = array_pop($element);
+  $service_endpoint = implode("/", $element);
+  $bucket = array_pop($element);
+  $file_name = basename($file_path);
+  $object_key = $prefix."/".$file_name;
+  
+  $s3Params = array(
+      'endpoint' => $service_endpoint,
+      'region'  => 'us-east-1',
+      'version' => '2006-03-01',
+      'credentials' => [
+        'key'    => 'dummy',
+        'secret' => 'dummy'
+      ]
+  );
+  // Create an S3Client
+  $s3Client = new S3Client($s3Params);
+
+  $source = fopen($file_path, 'rb');
+  $uploader = new ObjectUploader(
+    $s3Client,
+    $bucket,
+    $object_key,
+    $source
+  );
+
+  do {
+    try {
+      $result = $uploader->upload();
+    } catch (Exception $ex) {
+      $messages = array(
+        'message' => t('S3Client and Uploader error: @exception'),
+        'message_sub' => array('@exception' => $ex->getMessage()),
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+      watchdog_exception('islandora_video_streaming_panopto', $ex);
+    }
+  } while (!isset($result));
+}
+
+/*
+ * Creates manifest file, in temporary storage, needed for Panopto Video Upload
+ * @param $file_path
+ * @param $manifest_file_name
+ */
+function islandora_video_streaming_panopto_create_manifest_for_video($file_path = null, $manifest_file_name=null) {
+  $file_name = basename($file_path);
+
+  $template = file_get_contents('https://raw.githubusercontent.com/Panopto/upload-python-sample/master/simplest/upload_manifest_template.xml');
```

Comment:
Rather than depend on a fetch from GitHub, copy this resource locally into our code.  This could be a file added to the repo, or the contents of the XML embedded inline in the code.  This reduces external dependencies (network availability, resource availability, resource consistency).

Reply:
Done, will check for errors.

Reply:
Pending some fixes.